File: lib/store.ts
import path from "node:path";
import fs from "node:fs/promises";
import type { UIMessage } from "ai";

const DATA_DIR = path.resolve(process.cwd(), ".data");
const CHATS_DIR = path.join(DATA_DIR, "chats");
const INDEX = path.join(DATA_DIR, "conversations.json");

export interface ConversationMeta {
  id: string;
  title: string;
  createdAt: string;
  updatedAt: string;
  messageCount: number;
}

export interface Conversation extends ConversationMeta {
  messages: UIMessage[];
}

async function ensureDirs() {
  await fs.mkdir(CHATS_DIR, { recursive: true });
}

async function readIndex(): Promise<ConversationMeta[]> {
  try {
    return JSON.parse(await fs.readFile(INDEX, "utf8"));
  } catch {
    return [];
  }
}

async function writeIndex(list: ConversationMeta[]) {
  await ensureDirs();
  await fs.writeFile(INDEX, JSON.stringify(list, null, 2), "utf8");
}

function deriveTitle(messages: UIMessage[]): string {
  const firstUser = messages.find((m) => m.role === "user");
  const text =
    firstUser?.parts
      .filter((p) => p.type === "text")
      .map((p) => (p as { text: string }).text)
      .join(" ")
      .trim() ?? "";
  if (!text) return "New conversation";
  return text.length > 48 ? `${text.slice(0, 48)}…` : text;
}

export async function listConversations(): Promise<ConversationMeta[]> {
  return (await readIndex()).sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
}

export async function getConversation(id: string): Promise<Conversation | null> {
  try {
    return JSON.parse(await fs.readFile(path.join(CHATS_DIR, `${id}.json`), "utf8"));
  } catch {
    return null;
  }
}

export async function saveConversation(
  id: string,
  messages: UIMessage[],
  title?: string,
): Promise<ConversationMeta> {
  await ensureDirs();
  const now = new Date().toISOString();
  const index = await readIndex();
  const existing = index.find((c) => c.id === id);

  const meta: ConversationMeta = {
    id,
    title: title ?? existing?.title ?? deriveTitle(messages),
    createdAt: existing?.createdAt ?? now,
    updatedAt: now,
    messageCount: messages.length,
  };

  const conversation: Conversation = { ...meta, messages };
  await fs.writeFile(
    path.join(CHATS_DIR, `${id}.json`),
    JSON.stringify(conversation, null, 2),
    "utf8",
  );
  await writeIndex([meta, ...index.filter((c) => c.id !== id)]);
  return meta;
}

export async function renameConversation(id: string, title: string) {
  const index = await readIndex();
  const meta = index.find((c) => c.id === id);
  if (!meta) return;
  meta.title = title;
  meta.updatedAt = new Date().toISOString();
  await writeIndex(index);
  const conv = await getConversation(id);
  if (conv) {
    conv.title = title;
    await fs.writeFile(path.join(CHATS_DIR, `${id}.json`), JSON.stringify(conv, null, 2), "utf8");
  }
}

export interface SearchHit {
  id: string;
  title: string;
  snippet: string;
}

function plain(m: UIMessage): string {
  return (m.parts as { type: string; text?: string }[])
    .filter((p) => p.type === "text")
    .map((p) => p.text ?? "")
    .join(" ");
}

/** Search conversation titles and message text. */
export async function searchConversations(query: string): Promise<SearchHit[]> {
  const q = query.trim().toLowerCase();
  if (q.length < 2) return [];
  const index = await listConversations();
  const hits: SearchHit[] = [];
  for (const meta of index) {
    const conv = await getConversation(meta.id);
    if (!conv) continue;
    const body = conv.messages.map(plain).join("  ");
    const hay = `${meta.title}  ${body}`.toLowerCase();
    const at = hay.indexOf(q);
    if (at < 0) continue;
    const start = Math.max(0, at - 40);
    const snippet = `${start > 0 ? "…" : ""}${`${meta.title}  ${body}`.slice(start, at + q.length + 60).trim()}…`;
    hits.push({ id: meta.id, title: meta.title, snippet });
  }
  return hits;
}

export async function deleteConversation(id: string) {
  const index = await readIndex();
  await writeIndex(index.filter((c) => c.id !== id));
  await fs.rm(path.join(CHATS_DIR, `${id}.json`), { force: true });
}

File: backend/app/store.py
"""Conversation persistence and search (port of lib/store.ts). JSON files on disk."""

import json
import os
from datetime import datetime, timezone

from .config import settings


def _chats_dir() -> str:
    return os.path.join(settings.data_dir, "chats")


def _index_path() -> str:
    return os.path.join(settings.data_dir, "conversations.json")


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _read_index() -> list[dict]:
    try:
        with open(_index_path(), encoding="utf-8") as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError):
        return []


def _write_index(items: list[dict]) -> None:
    os.makedirs(settings.data_dir, exist_ok=True)
    with open(_index_path(), "w", encoding="utf-8") as f:
        json.dump(items, f, ensure_ascii=False, indent=2)


def _plain(message: dict) -> str:
    return " ".join(
        p.get("text", "") for p in message.get("parts", []) if p.get("type") == "text"
    )


def _derive_title(messages: list[dict]) -> str:
    first = next((m for m in messages if m.get("role") == "user"), None)
    text = _plain(first).strip() if first else ""
    if not text:
        return "New conversation"
    return text[:48] + "…" if len(text) > 48 else text


def list_conversations() -> list[dict]:
    return sorted(_read_index(), key=lambda c: c.get("updatedAt", ""), reverse=True)


def get_conversation(conv_id: str) -> dict | None:
    try:
        with open(os.path.join(_chats_dir(), f"{conv_id}.json"), encoding="utf-8") as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError):
        return None


def save_conversation(conv_id: str, messages: list[dict], title: str | None = None) -> dict:
    os.makedirs(_chats_dir(), exist_ok=True)
    index = _read_index()
    existing = next((c for c in index if c["id"] == conv_id), None)
    meta = {
        "id": conv_id,
        "title": title or (existing or {}).get("title") or _derive_title(messages),
        "createdAt": (existing or {}).get("createdAt") or _now(),
        "updatedAt": _now(),
        "messageCount": len(messages),
    }
    with open(os.path.join(_chats_dir(), f"{conv_id}.json"), "w", encoding="utf-8") as f:
        json.dump({**meta, "messages": messages}, f, ensure_ascii=False, indent=2)
    _write_index([meta] + [c for c in index if c["id"] != conv_id])
    return meta


def rename_conversation(conv_id: str, title: str) -> None:
    index = _read_index()
    meta = next((c for c in index if c["id"] == conv_id), None)
    if not meta:
        return
    meta["title"] = title
    meta["updatedAt"] = _now()
    _write_index(index)
    conv = get_conversation(conv_id)
    if conv:
        conv["title"] = title
        with open(os.path.join(_chats_dir(), f"{conv_id}.json"), "w", encoding="utf-8") as f:
            json.dump(conv, f, ensure_ascii=False, indent=2)


def delete_conversation(conv_id: str) -> None:
    _write_index([c for c in _read_index() if c["id"] != conv_id])
    try:
        os.remove(os.path.join(_chats_dir(), f"{conv_id}.json"))
    except OSError:
        pass


def search_conversations(query: str) -> list[dict]:
    q = query.strip().lower()
    if len(q) < 2:
        return []
    hits: list[dict] = []
    for meta in list_conversations():
        conv = get_conversation(meta["id"])
        if not conv:
            continue
        body = "  ".join(_plain(m) for m in conv.get("messages", []))
        hay = f"{meta['title']}  {body}".lower()
        at = hay.find(q)
        if at < 0:
            continue
        start = max(0, at - 40)
        snippet = f"{meta['title']}  {body}"[start : at + len(q) + 60].strip()
        hits.append({"id": meta["id"], "title": meta["title"], "snippet": snippet + "…"})
    return hits
